README.md
# Pi Sonar Node

Sonar Sensor Handling based on `pigpio` library.

To build:

```
cd ~/catkin_ws/src
git clone https://github.com/UbiquityRobotics/pi_sonar.git
cd ..
catkin_make
source devel/setup.bash
```

The node needs to access the pigpio daemon, hence the following steps to have it run at startup (if it does not do so already):

```
wget https://raw.githubusercontent.com/joan2937/pigpio/master/util/pigpiod.service
sudo cp pigpiod.service /etc/systemd/system
sudo systemctl enable pigpiod.service
sudo systemctl start pigpiod.service
```

To run:

```
roslaunch pi_sonar ubiquity_sonar.launch
```

## For developers only
Update pigpio:

git subtree pull --squash --prefix=pigpio https://github.com/joan2937/pigpio.git {ref to pull}


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package pi_sonar
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.6.4 (2021-07-19)
------------------
* Added boost format header to solve compile issues on 20.04 (`#17 <https://github.com/UbiquityRobotics/pi_sonar/issues/17>`_)
* removed pigpio, reworked code to work with pigpiod
* Contributors: MoffKalast, Rohan Agrawal

0.6.3 (2020-05-31)
------------------
* Added dependancy on libpigpio for build and run
* Contributors: Rohan Agrawal

0.6.2 (2020-05-31)
------------------
* Added dependancy on libpigpio
* Contributors: Rohan Agrawal

0.6.1 (2020-05-31)
------------------
* Added dependancy on libpigpiod-if
* Contributors: Rohan Agrawal

0.6.0 (2020-05-31)
------------------
* Removed pigpio, reworked code to work with pigpiod (`#12 <https://github.com/UbiquityRobotics/pi_sonar/issues/12>`_)
* Contributors: Rohan Agrawal, MoffKalast

0.5.0 (2020-05-30)
------------------
* Updating pigpio to v71 (`#11 <https://github.com/UbiquityRobotics/pi_sonar/issues/11>`_)
* Contributors: Jim Vaughan, MoffKalast, Rohan Agrawal

0.4.0 (2019-02-16)
------------------
* fix incorrect launch file in the README
* Basic diagnostics added
* Contributors: Rohan Agrawal, Rohit Suri

0.3.1 (2018-06-04)
------------------
* explictly add rpath
* Contributors: Rohan Agrawal

0.3.0 (2018-06-04)
------------------
* Initial Release
* Contributors: Jim Vaughan Rohan Agrawal


src/RangeMessage.h
/*
 * SonarRange.h
 *
 *  Created on: Feb 12, 2014
 *      Author: kurt
 */

#ifndef RANGEMESSAGE_H_
#define RANGEMESSAGE_H_

#include <stdint.h>
#include <ros/ros.h>
#include <std_msgs/Header.h>
#include <sensor_msgs/Range.h>

namespace ur {

class RangeMessage {
private:
	int sonar_id;
    sensor_msgs::Range msg;

public:
	RangeMessage();
	virtual ~RangeMessage();

	int getSonarIdentity();
	void setSonarIdentity(int id);
	int getRadiationType();
	void setRadiationType(int i);
	float getFieldOfView();
	void setFieldOfView(float fov);
	float getMinimumRange();
	void setMinimumRange(float min);
	float getMaximumRange();
	void setMaximumRange(float max);
	float getRange();
	void setRange(float r);
	void publish(ros::Publisher topic, int counter);
};

} /* namespace ur */

#endif /* RANGEMESSAGE_H_ */


src/RangeMessage.cpp
/*
 * RangeMessage.cpp
 *
 *  Created on: Feb 12, 2014
 *      Author: kurt
 */

#include "RangeMessage.h"

namespace ur {

RangeMessage::RangeMessage() {
	sonar_id = 999;                   /* unset */
	msg.radiation_type = 0;			/* ULTRASOUND=0; INFRARED=1 */
	msg.field_of_view = 0.175;		/* 10 degrees * pi / 180 = ~.175 radians*/
	msg.min_range = 0.005;				/* in meters */
	msg.max_range = 5;				/* in meters */
	msg.range = 0;                	/* default */
}

RangeMessage::~RangeMessage() {
	// TODO Auto-generated destructor stub
}

int RangeMessage::getSonarIdentity()
{
	return (sonar_id);
}
void RangeMessage::setSonarIdentity(int id)
{
	sonar_id = id;
}

int RangeMessage::getRadiationType()
{
	return (msg.radiation_type);
}

void RangeMessage::setRadiationType(int i)
{
	msg.radiation_type = i;
}

float RangeMessage::getFieldOfView()
{
    return (msg.field_of_view);
}

void RangeMessage::setFieldOfView(float fov)
{
	msg.field_of_view = fov;
}

float RangeMessage::getMinimumRange()
{
	return (msg.min_range);
}

void RangeMessage::setMinimumRange(float min)
{
	msg.min_range = min;
}

float RangeMessage::getMaximumRange()
{
	return (msg.max_range);
}

void RangeMessage::setMaximumRange(float max)
{
	msg.max_range = max;
}

float RangeMessage::getRange()
{
	return msg.range;
}

void RangeMessage::setRange(float r)
{
	msg.range = r;
}

void RangeMessage::publish(ros::Publisher topic, int counter)
{

	  switch (sonar_id)
	  {
	     case 1 :
	    	 msg.header.frame_id = "hercules_skirt1";
	    	 break;
	     case 2 :
	    	 msg.header.frame_id = "hercules_skirt2";
	    	 break;
	     case 3 :
	    	 msg.header.frame_id = "hercules_skirt3";
	    	 break;
	     case 4 :
	    	 msg.header.frame_id = "hercules_skirt4";
	    	 break;
	     case 5 :
	    	 msg.header.frame_id = "hercules_skirt5";
	    	 break;
	     case 6 :
	    	 msg.header.frame_id = "hercules_skirt6";
	    	 break;
	     case 7 :
	    	 msg.header.frame_id = "hercules_skirt7";
	    	 break;
	     case 8 :
	    	 msg.header.frame_id = "hercules_skirt8";
	    	 break;
	     case 9 :
	    	 msg.header.frame_id = "hercules_skirt9";
	    	 break;
	     case 10 :
	    	 msg.header.frame_id = "hercules_skirt10";
	    	 break;
	     default :
	    	 break;
	  }
		 msg.header.stamp = ros::Time::now();
    	 msg.header.seq = counter;
    	 topic.publish(msg);

}

} /* namespace ur */


src/ubiquity_sonar.cpp
/*
 * Copyright (c) 2018, Ubiquity Robotics
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing official
 * policies, either expressed or implied, of the FreeBSD Project.
 *
 */

#ifdef __arm__

#include <stdio.h>
#include <unistd.h>
#include <pigpio.h>

#include <ros/ros.h>
#include <sensor_msgs/Range.h>

#include <boost/algorithm/string.hpp>


class Sonar {
public:
  // BCM GPIO pins
  int trigger_pin;
  int echo_pin;

  uint32_t start_tick;
  uint32_t elapsed_ticks;

  std::string frame;
  ros::Publisher pub;

  Sonar(int trigger_pin, int echo_pin, int id, ros::NodeHandle& nh)
  {
    this->trigger_pin = trigger_pin;
    this->echo_pin = echo_pin;

    start_tick = 0;
    elapsed_ticks = 0;

    frame = str(boost::format{"sonar_%1%"} % id);
    pub = nh.advertise<sensor_msgs::Range>(frame, 1);
  }
};


static std::vector<Sonar> sonars;


/* Trigger the next sonar */
void sonar_trigger()
{
   static int sonar = 0;

   int pin = sonars[sonar].trigger_pin;
   gpioWrite(pin, PI_ON);
   gpioDelay(10); /* 10us trigger pulse */
   gpioWrite(pin, PI_OFF);

   sonar++;
   if (sonar >= sonars.size()) {
     sonar = 0;
   }
}

/* Handle pin change */
void echo_callback(int pin, int level, uint32_t tick)
{
  for (auto& sonar : sonars) {
    if (pin == sonar.echo_pin) {
      if (level == PI_ON) {
        sonar.start_tick = tick;
      }
      else if (level == PI_OFF) {
        uint32_t elapsed = tick - sonar.start_tick;
        sonar.elapsed_ticks = elapsed;
        //printf("sonar %d elapsed %d\n", sonar, elapsed);
      }
      return;
     }
   }
   printf("Unexpected GPIO pin event, pin %d\n", pin);
}

int setup_gpio()
{
  if (gpioInitialise() < 0) {
    return false;
  }

  for (const auto& sonar : sonars) {
    gpioSetMode(sonar.trigger_pin, PI_OUTPUT);
    gpioWrite (sonar.trigger_pin, PI_OFF);

    gpioSetMode(sonar.echo_pin, PI_INPUT);

    /* monitor sonar echos */
    gpioSetAlertFunc(sonar.echo_pin, echo_callback);
  }

  /* update sonar 20 times a second, timer #0 */
  gpioSetTimerFunc(0, 50, sonar_trigger); /* every 50ms */

  return true;
}

int main(int argc, char *argv[])
{
  ros::init(argc, argv, "ubiquity_sonar");
  ros::NodeHandle nh("~");

  double field_of_view;
  double min_range;
  double max_range;

  nh.param<double>("field_of_view", field_of_view, .43632347);
  nh.param<double>("min_range", min_range, .05);
  nh.param<double>("max_range", max_range, 10);

  ros::Publisher pub = nh.advertise<sensor_msgs::Range>("/sonars", 5);

  // pin numbers are specific to the hardware
  sonars.push_back(Sonar(20, 21, 0, nh));
  sonars.push_back(Sonar(12, 16, 1, nh));
  sonars.push_back(Sonar(23, 24, 2, nh));
  sonars.push_back(Sonar(27, 22, 3, nh));
  sonars.push_back(Sonar(19, 26, 4, nh));

  if (!setup_gpio()) {
    ROS_ERROR("Cannot initalize gpio");
    return 1;
  }

  ros::Rate rate(50);

  sensor_msgs::Range msg;
  msg.field_of_view = field_of_view;
  msg.min_range = min_range;
  msg.max_range = max_range;
  msg.radiation_type = sensor_msgs::Range::ULTRASOUND;

  ROS_INFO("Ubiquity Sonar node ready");

  while (ros::ok()) {
    for (auto& sonar : sonars) {
      uint32_t elapsed_ticks = sonar.elapsed_ticks;
      if (elapsed_ticks != 0) {
        // clear so we don't publish again
        sonar.elapsed_ticks = 0;
        // seconds = ticks / 1000000
        // speed of sound = 343 m/s
        // for round trip, halve the distance
        msg.range = (float)elapsed_ticks * 0.0001715;
        msg.header.stamp = ros::Time::now();
        msg.header.frame_id = sonar.frame;
        pub.publish(msg);
        sonar.pub.publish(msg);
      }
    }
    rate.sleep();
  }

  gpioTerminate();

  return 0;
}

#else

#include <stdio.h>

int main(int argc, char **argv) {
  fprintf(stderr, "ubiquity_sonar only works on the Raspberry Pi\n");
  return 1;
}

#endif // __arm__


src/pi_sonar.cpp
/*
 * Copyright (c) 2018, Ubiquity Robotics
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing official
 * policies, either expressed or implied, of the FreeBSD Project.
 *
 */

#ifdef __arm__

#include <stdio.h>
#include <unistd.h>
#include <pigpiod_if2.h>

#include <ros/ros.h>
#include <sensor_msgs/Range.h>

#include <boost/algorithm/string.hpp>
#include <boost/format.hpp>

#include <diagnostic_updater/diagnostic_updater.h>
#include <diagnostic_updater/publisher.h>

double min_freq = 0.5;
double max_freq = 60;

int gpio = -1;
pthread_t* sonarthread;

class Sonar {
public:
    // BCM GPIO pins
    int trigger_pin;
    int echo_pin;
    int id;

    uint32_t start_tick;
    uint32_t elapsed_ticks;

    std::string frame;
    ros::Publisher pub;

    bool range_error = false;

    std::unique_ptr<diagnostic_updater::Updater> updater;
    std::unique_ptr<diagnostic_updater::HeaderlessTopicDiagnostic> pub_freq;
 
    Sonar(int trigger_pin, int echo_pin, int id, ros::NodeHandle& nh)
    {
        this->trigger_pin = trigger_pin;
        this->echo_pin = echo_pin;
        this->id = id;

        start_tick = 0;
        elapsed_ticks = 0;

        frame = str(boost::format{"sonar_%1%"} % id);
        pub = nh.advertise<sensor_msgs::Range>(frame, 1);
        
        updater.reset(new diagnostic_updater::Updater());
        updater->setHardwareIDf("sonar_%d",id);
        pub_freq.reset(new diagnostic_updater::HeaderlessTopicDiagnostic(
            str(boost::format{"/pi_sonar/sonar_%1%"} % id), 
            *updater,
            diagnostic_updater::FrequencyStatusParam(&min_freq, &max_freq, 0.1, 10))
        );
        updater->add(str(boost::format{"Sonar %1% Range Checker"} % id),
            this,
            &Sonar::range_check
        );
    }

    void range_check(diagnostic_updater::DiagnosticStatusWrapper &stat)
    {
        if (range_error) {
            stat.summary(diagnostic_msgs::DiagnosticStatus::WARN, "Range out of bounds!");
        }
        else {
            stat.summary(diagnostic_msgs::DiagnosticStatus::OK, "Range within bounds!");
        }
    }
};


static std::vector<Sonar> sonars;

/* Trigger the next sonar */
void sonar_trigger()
{
     static int sonar = 0;

     int pin = sonars[sonar].trigger_pin;
     gpio_write(gpio, pin, PI_ON);
     
     int waittime = get_current_tick(gpio);
     while(get_current_tick(gpio) - waittime < 10){
        /* wait for 10us trigger pulse */
     }

     gpio_write(gpio, pin, PI_OFF);

     sonar++;
     if (sonar >= sonars.size()) {
         sonar = 0;
     }
}

/* Sonar pulsing thread */ 
void* sonar_thread(void* data) 
{
    /* every 50ms, with probably garbage accuracy */
    while (1){
        time_sleep(0.05);
        sonar_trigger();
    }
    return NULL;
}


/* Handle pin change */
void echo_callback(int pigpio, uint32_t pin, uint32_t level, uint32_t tick)
{
    for (auto& sonar : sonars) {
        if (pin == sonar.echo_pin) {
            if (level == PI_ON) {
                sonar.start_tick = tick;
            }
            else if (level == PI_OFF) {
                uint32_t elapsed = tick - sonar.start_tick;
                sonar.elapsed_ticks = elapsed;
                //printf("sonar %d elapsed %d\n", sonar, elapsed);
            }
            return;
         }
     }
     printf("Unexpected GPIO pin event, pin %d\n", pin);
}

int setup_gpio()
{
    gpio = pigpio_start(NULL, NULL);
    if (gpio < 0) {
        return false;
    }

    for (const auto& sonar : sonars) {
        set_mode(gpio, sonar.trigger_pin, PI_OUTPUT);
        gpio_write(gpio, sonar.trigger_pin, PI_OFF);

        set_mode(gpio, sonar.echo_pin, PI_INPUT);

        /* monitor sonar echos */
        callback(gpio, sonar.echo_pin, EITHER_EDGE, echo_callback);
    }

    /* update sonar 20 times a second, timer #0 */
    void* extra_param_because_reasons = 0;
    sonarthread = start_thread(sonar_thread, extra_param_because_reasons);

    return true;
}

int main(int argc, char *argv[])
{
    ros::init(argc, argv, "pi_sonar");
    ros::NodeHandle nh("~");

    double field_of_view;
    double min_range;
    double max_range;

    nh.param<double>("field_of_view", field_of_view, 0.43632347);
    nh.param<double>("min_range", min_range, 0.05);
    nh.param<double>("max_range", max_range, 10);

    ros::Publisher pub = nh.advertise<sensor_msgs::Range>("/sonars", 5);

    // pin numbers are specific to the hardware
    sonars.push_back(Sonar(20, 21, 0, nh));
    sonars.push_back(Sonar(12, 16, 1, nh));
    sonars.push_back(Sonar(23, 24, 2, nh));
    sonars.push_back(Sonar(27, 22, 3, nh));
    sonars.push_back(Sonar(19, 26, 4, nh));

    if (!setup_gpio()) {
        ROS_ERROR("Cannot initalize gpio");
        return 1;
    }

    ros::Rate rate(50);

    sensor_msgs::Range msg;
    msg.field_of_view = field_of_view;
    msg.min_range = min_range;
    msg.max_range = max_range;
    msg.radiation_type = sensor_msgs::Range::ULTRASOUND;

    ROS_INFO("Pi Sonar node ready");

    while (ros::ok()) {
        for (auto& sonar: sonars) {
            uint32_t elapsed_ticks = sonar.elapsed_ticks;
            if (elapsed_ticks != 0) {
                // clear so we don't publish again
                sonar.elapsed_ticks = 0;
                // seconds = ticks / 1000000
                // speed of sound = 343 m/s
                // for round trip, halve the distance
                msg.range = (float)elapsed_ticks * 0.0001715;
                sonar.range_error = (msg.range < min_range || msg.range > max_range);
                msg.header.stamp = ros::Time::now();
                msg.header.frame_id = sonar.frame;
                pub.publish(msg);
                sonar.pub.publish(msg);
                sonar.pub_freq->tick();
            }
            sonar.updater->update();
        }
        rate.sleep();
    }

    stop_thread(sonarthread);
    pigpio_stop(gpio);

    return 0;
}

#else

#include <stdio.h>

int main(int argc, char **argv) {
    fprintf(stderr, "pi_sonar only works on the Raspberry Pi\n");
    return 1;
}

#endif // __arm__
